docs: add redirect and update updating-to-v11 to v12 (#41625)
angular/angular: aio/content/guide/updating-to-version-12.md

# Updating Angular

This guide contains information related to updating to Angular version 12.

## Updating CLI applications

For step-by-step instructions on how to update to the latest Angular release and leverage our automated migration tools to do so, use the interactive update guide at [update.angular.io](https://update.angular.io).

## Building applications with Ivy

> For libraries, View Engine will be deprecated and removed in version 13.

Reviewer: ```suggestion
For libraries, View Engine is deprecated and will be removed in version 13.
```

> New libraries built with version 12 default to Ivy.

Reviewer: ```suggestion
New libraries created with version 12 or later default to Ivy.
```

> 

Reviewer: I would include a link here to the documentation for `ng-linker` (when it's done) with something like
"More more information about distributing libraries with View Engine and Ivy, see ..."

## Changes and deprecations in version 12

<div class="alert is-helpful">

   For information about Angular's deprecation and removal practices, see [Angular Release Practices](guide/releases#deprecation-practices "Angular Release Practices: Deprecation practices").

</div>

* Applications can no longer build with View Engine by setting `enableIvy: false`.
  Support for building libraries with View Engine, for backwards compatibility, is deprecated and will be removed in Angular version 13.
  New libraries created with Angular version 12 will default to building and distributing with Ivy.
  For more information, see [Creating Libraries](guide/creating-libraries).

> * The Ivy-based language service is now on by default.

Reviewer: ```suggestion
* The Ivy-based IDE language service is now on by default.
```
I would probably also make "language service" a link here to more info

  See [PR #1279](https://github.com/angular/vscode-ng-language-service/pull/1279).

> * The previous message ID generation algorithm is deprecated.

Reviewer: ```suggestion
* Angular's View Engine-based algorithm for generating i18n message IDs is deprecated.
```
@petebacondarwin or @crisbeto is this proposed edit accurate? I want to cut out any ambiguity about which specific behavior is deprecated and how the new behavior is enabled.

Author: I've added this in the interim as it is at least clearer than what I had.

  Angular version 12 adds a new flag to `localize-extract` called `--migrateMapFile` which  generates a JSON file that can be used to map legacy message IDs to canonical ones.
  There is also a new script called `localize-migrate` that can use the mapping file which `localize-extract` generates and migrate all of the IDs in the files that are passed in.
  For better stability, if you are using Angular's `i18n`, run this migration to move to the new message ID generation algorithm.

>   If you don't run this migration, all your generated message IDs will  change when Angular removes the View Engine compiler.

Reviewer: ```suggestion
  If you don't run this migration, all your generated message IDs will change when Angular removes the View Engine compiler.
```
extra space

  To update web worker usage, where `./app.worker` is the actual worker name, change `new Worker('./app.worker', ...)` to `new Worker(new URL('./app.worker', import.meta.url), ...)`.
  See [PR #20466](https://github.com/angular/angular-cli/pull/20466).
* Critical CSS inlining is now enabled by default.
  To turn this off, set `inlineCritical` to false.
  See [PR #20096](https://github.com/angular/angular-cli/pull/20096) and the [Style preprocessor options](guide/workspace-config#optimization-configuration) section of [Angular workspace configuration](guide/workspace-config).
* `ng build` now produces production bundle by default.
  See [PR #20128](https://github.com/angular/angular-cli/pull/20128).
* Previously, the Forms module ignored `min` and `max` attributes defined on the `<input type="number">`.
  Now these attributes trigger `min` and `max` validation logic in cases where `formControl`, `formControlName`, or `ngModel` directives are also present on a given input.
  See [PR #39063](https://github.com/angular/angular/pull/39063).

{@a deprecations}
### New deprecations

> * Support for Internet Explorer 11 has been deprecated.

Reviewer: ```suggestion
* Support for Internet Explorer 11 is deprecated.
```

  See [Deprecated APIs and features](guide/deprecations) and [Microsoft 365 apps say farewell to Internet Explorer 11 and Windows 10 sunsets Microsoft Edge Legacy](https://techcommunity.microsoft.com/t5/microsoft-365-blog/microsoft-365-apps-say-farewell-to-internet-explorer-11-and/ba-p/1591666).
* Sass imports from `@angular/material/theming` are deprecated. There is a new Angular Material Sass API for `@use`.
  Run the migration script `ng g @angular/material:themingApi` to switch all your Sass imports for Angular CDK and Angular Material to the new API and `@use`.
* Support for publishing libraries with View Engine has been deprecated:
  - You can now compile libraries in [_partial_ compilation mode](https://angular.io/guide/angular-compiler-options#compilationmode) to generate Ivy compatible output that will be _linked_ when an application using that library is bundled.

>   - New libraries you create with the Angular CLI default to partial compilation mode, and do not support View Engine. You can still build a library with View Engine. See [Creating libraries](guide/creating-libraries) for more information.

Author: I have just such a thing cooking. May I interest you in https://github.com/angular/angular/pull/41485?  :stew:

  See [PR #20514](https://github.com/angular/angular-cli/pull/20514).
* For new applications, strict mode is now the default in the CLI.
  See [PR #20029](https://github.com/angular/angular-cli/pull/20029).
* Add `emitEvent` option for `AbstractControl` class methods.
  See [PR #31031](https://github.com/angular/angular/pull/31031).
* Support` APP_INITIALIZER` to work with observables.
  See [PR #31031](https://github.com/angular/angular/pull/31031).
* `HttpClient` supports specifying request metadata.
  See [PR #25751](https://github.com/angular/angular/pull/25751).

{@a breaking-changes}
### Breaking changes in Angular version 12

* Add support for TypeScript 4.2.
  TypeScript 4 is no longer supported.
  The supported range of TypeScript versions is 4.2.3 to 4.2.x.
  See [PR #41158](https://github.com/angular/angular/pull/41158).
* Angular CDK and Angular Material internally now use the new [Sass module system](https://sass-lang.com/blog/the-module-system-is-launched), which is actively maintained by the Sass team at Google.
  Consequently, applications can no longer consume Angular CDK/Material's Sass with the `node-sass` npm package.

>   `node-sass` is unmaintained and does not support newer Sass features. Instead, applications must use the `sass` npm package, or the `sass-embedded` npm package for the `sass-embedded` beta.

Author: I added the links (thank you!!) but didn't find anything about what "`sass-embedded` beta" means. I found [this repo](https://github.com/sass/embedded-protocol) but is that what you're thinking? I also found [this tweet](https://twitter.com/SassCSS/status/1383151350851072003), but nothing like an article or blog post.

>   `node-sass` is unmaintained and does not support newer Sass features. Instead, applications must use the `sass` npm package, or the `sass-embedded` npm package for the `sass-embedded` beta.

Reviewer 2: I don't know much about it either, so feel free to leave it as is :grin:

> * The Angular tooling now uses Webpack 5 to build applications. Webpack 4 usage and support has been removed.

Author: Here, the word tooling is a noun, so it's what uses Webpack.

  You don't need to make any project level configuration changes to use the upgraded Webpack version when using the official Angular builders.
  Custom builders based on this package that use the experimental programmatic APIs may need to be updated to become compatible with Webpack 5.

>   See [PR #20466](https://github.com/angular/angular-cli/pull/20466).

Author: I can't reproduce this formatting issue. Is your screen zoomed in by chance? I've broken the layout by forgetting I zoomed in.

* Webpack 5 generates similar but differently named files for lazy-loaded JavaScript files in development configurations when the `namedChunks` option is enabled.
  For the majority of users this change should have no effect on the application or build process.
  Production builds should also not be affected as the `namedChunks` option is disabled by default in production configurations.
  However, if a project's post-build process makes assumptions as to the file names then adjustments may need to be made to account for the new naming paradigm.
  Such post-build processes could include custom file transformations after the build, integration into service-side frameworks, or deployment procedures.
  An example of a development file name change is `lazy-lazy-module.js` becoming `src_app_lazy_lazy_module_ts.js`.

>   See [PR #20466](https://github.com/angular/angular-cli/pull/20466).

Author: Same as above: https://github.com/angular/angular/pull/41625#discussion_r620659820

* Webpack 5 now includes web worker support.
  However, the structure of the URL within the worker constructor must be in a specific format that differs from the current requirement.
  To update Web worker usage, where `./app.worker` is the actual worker name, change `new Worker('./app.worker', ...)` to `new Worker(new URL('./app.worker', import.meta.url), ...)`.
  See [PR #20466](https://github.com/angular/angular-cli/pull/20466).
* Critical CSS inlining is now enabled by default.
  To turn this off, set `inlineCritical` to false.

>   See [Angular CLI PR #20096](https://github.com/angular/angular-cli/pull/20096) and the [Style preprocessor options](https://angular.io/guide/workspace-config#optimization-configuration) section of [Angular workspace configuration](https://angular.io/guide/workspace-config).

Author: Same as above: https://github.com/angular/angular/pull/41625#discussion_r620659820

Leaving open for reference.

* `ng build` now produces production bundle by default.
  See [Angular CLI PR #20128](https://github.com/angular/angular-cli/pull/20128).
* Previously, the Forms module ignored `min` and `max` attributes defined on the `<input type="number">`
  Now these attributes trigger `min` and `max` validation logic in cases where `formControl`, `formControlName` or `@ngModel()` directives are also present on a given input.
  See [PR #39063](https://github.com/angular/angular/pull/39063).

{@a deprecations}
### New Deprecations

> * Deprecate support for IE 11.

Reviewer 2: ```suggestion
* Support for IE 11 has been deprecated.
```

  See [Deprecated APIs and features](guide/deprecations) and [Microsoft 365 apps say farewell to Internet Explorer 11 and Windows 10 sunsets Microsoft Edge Legacy](https://techcommunity.microsoft.com/t5/microsoft-365-blog/microsoft-365-apps-say-farewell-to-internet-explorer-11-and/ba-p/1591666).
* There is a new Angular Material Sass API for `@use`.
  Run the migration script `ng g @angular/material:themingApi` to switch all your Sass imports for Angular CDK and Angular Material to the new API and `@use`.

>   Sass imports from `@angular/material/theming` are deprecated.

Reviewer 2: I would move this sentence to the beginning of the bullet-point to make clear what is being deprecated.

> * Support for publishing libraries with Ivy:

Reviewer 2: This sounds like publishing libraries with Ivy is deprecated. How about:

```suggestion
* Support for publishing libraries with View Engine has been deprecated:
```

  - You can now compile libraries in [_partial_ compilation mode](https://angular.io/guide/angular-compiler-options#compilationmode) to generate ivy compatible output that will be _linked_ when an application using that library is bundled.
  - New libraries you create with the Angular CLI default to partial compilation mode, and do not support View Engine.
  - Libraries compiled in partial compilation mode will not contain legacy `i18n` message ids.
  If the library was previously compiled by View Engine, and contained legacy `i18n` message ids, then applications may have translation files that you'll need to migrate to the new message id format.

>   <!-- For more information see, [](NEED LINK). -->

Reviewer 2: Can this be removed now?

```suggestion
```

>   - For context, see [Issue #38366](https://github.com/angular/angular/issues/38366).

Reviewer 2: I think it would be helpful to mention that one can still build a library with View Engine and link to some place (where?) that contains info on how to do it?

  See [PR #20514](https://github.com/angular/angular-cli/pull/20514).
* BREAKING CHANGE: Enable critical CSS inlining by default.
  See [Angular CLI PR #20096](https://github.com/angular/angular-cli/pull/20096).
* `ng build` now produces production bundle by default.
  See [Angular CLI PR #20128](https://github.com/angular/angular-cli/pull/20128).
* Enable strict mode by default in the CLI.
  See [Commit #](https://github.com/angular/angular-cli/commit/b105ed63c7610dd1397a3d24d4a7439564a019aa).
  <!-- OR? -->
  See [PR #20029](https://github.com/angular/angular-cli/pull/20029).
* Add min and max Forms validators.
  See [PR #39063](https://github.com/angular/angular/pull/39063).
* Add `emitEvent` option for `AbstractControl` class methods.
  See [PR #31031](https://github.com/angular/angular/pull/31031).
* Support APP_INITIALIZER to work with observables.
  See [PR #31031](https://github.com/angular/angular/pull/31031).
* `HttpClient` supports specifying request metadata.
  See [PR #25751](https://github.com/angular/angular/pull/25751).
* Linked library code will no longer contain any legacy `i18n` message ids.

> For this reason, libraries that existed before version 12 that contained `i18n` template tags need to inform their clients that the message ids will have changed and instruct them on how to migrate them.

Author: @petebacondarwin what do you think of having this note you supplied in https://github.com/angular/angular/pull/41485#issuecomment-818745194 in this location?

Reviewer 3: Generally this appears to be the right place to put this message.

I think we should have a consistent way of talking about these concepts. In this doc we have `ng-linker` and "linked library", but we have used the phrase "partial" compilation, e.g. in the angular compiler options.

I would put the two bullet points that mention the new library format next to each other. Can we have sub-bullets? Perhaps it could be worded something like:

```
* Support for publishing libraries with Ivy:
  - Libraries can now be compiled in "partial" compilation mode to generate ivy compatible output that will be "linked" when an application using that library is bundled.
  - New libraries created by the Angular CLI default to "partial" compilation mode, and do not support View Engine.
  - Libraries compiled in "partial" compilation mode will not contain legacy `i18n` message ids. If the library was previously compiled by View Engine, and contained legacy `i18n` message ids, then applications may have translation files that need to be migrated to the new message id format.
```

We should also document, somewhere, how to do this migration (@crisbeto ?) and then link to that from here.

Reviewer 4: I talked to @aikidave about documenting the migration a few weeks ago, but I don't know how far along it is.

Reviewer 5: Documented in https://github.com/angular/angular/pull/41828

Author: Added link. Thanks, @aikidave.

# Updating Angular

This guide contains information related to updating to the latest version of Angular.

## Updating CLI Apps

For step-by-step instructions on how to update to the latest Angular release and leverage our automated migration tools to do so, use the interactive update guide at [update.angular.io](https://update.angular.io).

## Changes and Deprecations in Version 12

<div class="alert is-helpful">

   For information about Angular's deprecation and removal practices, see [Angular Release Practices](guide/releases#deprecation-practices "Angular Release Practices: Deprecation practices").

</div>

{@a breaking-changes}
### New Breaking changes in Angular version 12

<!-- Are all these breaking? If all are not, should we separate them into breaking and non-breaking?-->
* Deprecate support for IE 11.
  See [Deprecated APIs and features](guide/deprecations) and [Microsoft 365 apps say farewell to Internet Explorer 11 and Windows 10 sunsets Microsoft Edge Legacy](https://techcommunity.microsoft.com/t5/microsoft-365-blog/microsoft-365-apps-say-farewell-to-internet-explorer-11-and/ba-p/1591666).

> * Deprecate `ViewEngine`.

Author: Thanks so much for this, @gkalpak!

Hey @aikidave, can you point me in the right direction for confirming this?

> * Deprecate `ViewEngine`.

Author: resolved offline

  <!-- Is version 4 no longer supported? -->
* Set Ivy-based language service on by default.
  See [PR #1279](https://github.com/angular/vscode-ng-language-service/pull/1279).
* Migrate away from legacy message IDs.
  See [PR #41026](https://github.com/angular/angular/pull/41026).
* Migrate from the unmaintained `libsass` compiler implementation `node-sass` to the `sass` npm package, which is actively maintained by the Sass team at Google.
  Migrate to the `@use` syntax.
  See [Theming your Angular Material app](https://material.angular.io/guide/theming).
* CLI production-ready webpack 5 support.
  Remove usage and support of Webpack 4.
  <!-- Noting specifically here that these are breaking in case all changes aren't breaking. -->
  BREAKING CHANGE: Webpack 5 lazy loaded file name changes.
  BREAKING CHANGE: Webpack 5 web worker support.
  See [PR #20466](https://github.com/angular/angular-cli/pull/20466).
* Add build option `inlineStyleLanguage` that allows a project to define the stylesheet language used in an application's inline component styles.
  See [PR #20514](https://github.com/angular/angular-cli/pull/20514).

> * BREAKING CHANGE: Enable critical CSS inlining by default.

Author: Yes, so those are notes to me in that I understand that those are certainly breaking, my question is, are all the changes tentatively in this section breaking ([see line 20 comment](https://github.com/angular/angular/pull/41625/files#diff-f332eb8445f10de5c8ff582d53dad312a95a294eeede76444c43b04eff26214aR20))? 

If some are breaking and some aren't, then we should either specify which in the list are breaking (and adjust the header accordingly) or have the breaking ones in their own section. For example, the `inlineStyleLanguage` change above this one seems that it wouldn't be breaking, but I need confirmation. Of the list items that don't specify breaking, would you know which are breaking and which aren't, @gkalpak? If you know specific ones, I'll take that too!

> * BREAKING CHANGE: Enable critical CSS inlining by default.

Reviewer 2: I see. I can't be 100% sure from the descriptions, but there seems like there are a couple non-breaking changes there:
- ``Add build option `inlineStyleLanguage` ``
- `Add min and max Forms validators`
- `` `HttpClient` supports specifying request metadata``
- `Support for publishing libraries with Ivy`

Better to confirm with someone :grin:
If you have commits corresponding to each change, you can (theoretically) easily check whether the change is a breaking one by looking for a `BREAKING CHANGE` section in the commit message.

> * BREAKING CHANGE: Enable critical CSS inlining by default.

Author: Oh thank you, @gkalpak! I didn't know about the `BREAKING CHANGE` section in the commit message. I'll check them. 🌸

* Support for publishing libraries with Ivy:
  - You can now compile libraries in [_partial_ compilation mode](https://angular.io/guide/angular-compiler-options#compilationmode) to generate ivy compatible output that will be _linked_ when an application using that library is bundled.
  - New libraries you create with the Angular CLI default to partial compilation mode, and do not support View Engine.
  - Libraries compiled in partial compilation mode will not contain legacy `i18n` message ids.
  If the library was previously compiled by View Engine, and contained legacy `i18n` message ids, then applications may have translation files that you'll need to migrate to the new message id format.
  <!-- For more information see, [](NEED LINK). -->
  - For context, see [Issue #38366](https://github.com/angular/angular/issues/38366).

<!-- Do we have new items to add here? -->

{@a deprecations}
### New Deprecations

| Area                          | API or Feature                                     | May be removed in |
| ----------------------------- | -------------------------------------------------- | ----------------- |

> | `@angular/core/testing`       | Rename `async` to `waitForAsync`                       | <!--v11--> v13 |

Reviewer 2: ```suggestion
| `@angular/core/testing`       | Rename `async` to `waitForAsync`                   | <!--v11--> v13 |
```

Author: Hey @gkalpak, I'm not seeing the diff here. Also, does this need to be removed entirely from the doc as this is from the current doc? https://angular.io/guide/updating-to-version-11#new-deprecations

Reviewer 2: The difference is that there is some extra trailing space in the second column.

If this was present in the v11 doc, though, then it is not a new deprecation in v12, so it should be removed I guess.

<!-- Removals for v12? -->
{@a removals}
### New Removals of Deprecated APIs

The following APIs have been removed starting with version 11.0.0*:

| Package          | API            | Replacement | Notes |
| ---------------- | -------------- | ----------- | ----- |

> | `@angular/router`| `NavigationExtras#preserveQueryParams` | no action needed | NavigationExtras#preserveQueryParams has been removed from `@angular/router`.|

Reviewer 2: Nit:

```suggestion
| `@angular/router` | `NavigationExtras#preserveQueryParams` | no action needed | `NavigationExtras#preserveQueryParams` has been removed from `@angular/router`.|
```

Author: This is also from the current doc. Does this info stay? Are there cases where a section is here for multiple releases?

Reviewer 2: AFAICT, this docs should only contain deprecations/removals/breaking changes pertaining to the v12 release. So, if something was included in the v11 doc, it should probably not be in the v12 doc as well :grin:

> | `@angular/core` | `CollectionChangeRecord` | no action needed | CollectionChangeRecord has been removed from `@angular/core`.|

Reviewer 2: ```suggestion
| `@angular/core` | `CollectionChangeRecord` | no action needed | `CollectionChangeRecord` has been removed from `@angular/core`.|
```

Author: Same question as above:

> This is also from the current doc. Does this info stay? Are there cases where a section is here for multiple releases?

Author: resolved offline

| `@angular/core` | `ViewEncapsulation.Native` | no action needed | Angular previously supported a view encapsulation mode `ViewEncapsulaion.Native` that was based on the v0 Shadow DOM Draft APIs. These APIs have been superceeded by the final Shadow DOM APIs, which are enabled via `ViewEncapsulation.ShadowDom`. For background information about this change, see [Web Components update: more time to upgrade to v1 APIs](https://developers.google.com/web/updates/2019/07/web-components-time-to-upgrade).|

{@a ivy}

## Ivy features and compatibility

Since version 9, Angular Ivy is the default rendering engine. If you haven't heard of Ivy, you can read more about it in the [Angular Ivy guide](guide/ivy).

* Among other features, Ivy introduces more comprehensive type-checking within templates. For details, see [Template Type-checking](guide/template-typecheck).

* For general guidance on debugging and a list of minor changes associated with Ivy, see the [Ivy compatibility guide](guide/ivy-compatibility).

> * For help with opting out of Ivy, see the instructions [here](guide/ivy#opting-out-of-angular-ivy).

Reviewer 2: This might not be relevant for v12 apps (if it is not possible to build apps with `ViewEngine`).

Author: Same question as above:

> This is also from the current doc. Does this info stay? Are there cases where a section is here for multiple releases?

Author: resolved offline

Reviewer 2: What was the resolution?

Reviewer 2: NVM, I see this section was removed from the doc :+1:
